dask_expr/_merge.py

```diff
@@ -247,6 +254,63 @@ def _simplify_up(self, parent):
                     return type(parent)(result)
                 return result[parent_columns]
 
+    def _combine_similar(self, root: Expr):
+        # Push projections back up to avoid performing the same merge multiple times
+
+        def _remove_projections(frame):
+            operations, columns = [], []
+            frame_base = frame
+            while isinstance(
+                frame, (AssignPartitioningIndex, Shuffle, Filter, Projection)
+            ):
+                # Have to respect ops that were injected while lowering or filters
+                if isinstance(frame, Projection):
```

Comment:
A lot of this logic feels similar to `Blockwise._combine_similar`. Have you considered the possibility of generalizing that code, and using it in other places?